pkg/mimirpb/split_test.go

```diff
@@ -166,7 +166,7 @@ func TestSplitWriteRequestByMaxMarshalSize_WriteRequestHasChanged(t *testing.T)
 
 	// If the fields of WriteRequest have changed, then you will probably need to modify
 	// the SplitWriteRequestByMaxMarshalSize() implementation accordingly!
-	assert.ElementsMatch(t, []string{"Timeseries", "Source", "Metadata", "SkipLabelValidation", "skipUnmarshalingExemplars"}, fieldNames)
+	assert.ElementsMatch(t, []string{"Timeseries", "Source", "Metadata", "SkipLabelValidation", "skipUnmarshalingExemplars", "buffer"}, fieldNames)
```

Comment:
I think the most correct thing to do in `SplitWriteRequestByMaxMarshalSize()` is to retain a reference of the original `buffer` in each split `WriteRequest` (and by calling `bufferRef()` for each split `WriteRequest`). We could also mention it in the `SplitWriteRequestByMaxMarshalSize()` function doc.

Then where we call `SplitWriteRequestByMaxMarshalSize()` (which is `marshalWriteRequestToRecords()`) we should call the `FreeBuffer()` on each split request once we've done marshalling it.

The only annoying thing of this approach is that `SplitWriteRequestByMaxMarshalSize()` will just return the input `WriteRequest` is not splitting happened at all. We could slightly modify it to also return a `bool` to indicate if splitting happened, so that the caller will call `FreeBuffer()` only if splitting happened.

Reply:
I'm not sure I agree on this. These are _internal things we do with the WriteRequest_, they shouldn't affect the outer lifecycle: we had a WriteRequest, we call FreeBuffer after using it (and _using it_ here implies splitting it into multiple ones).

Maybe we should clarify in the comment of `SplitWriteRequestByMaxMarshalSize` that the requests returned in the slice _might_ still retain references to the original WriteRequest.

Reply:
I agree with you Oleg, and suggested adding the comment instead on internal slack thread as well.

Reply:
After reviewing the code, I agree w/ @colega and @pstibrany and opted to document that the `WriteRequests` returned by `SplitWriteRequestByMaxMarshalSize` retain references to fields in the original `WriteRequest`.`